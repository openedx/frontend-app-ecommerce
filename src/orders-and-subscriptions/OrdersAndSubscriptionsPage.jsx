import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getConfig } from '@edx/frontend-platform';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';
import { FormattedMessage, useIntl } from '@edx/frontend-platform/i18n';

import { BasicAlert, PageLoading } from '../components';

import Subscriptions, { fetchSubscriptions } from '../subscriptions';
import OrderHistory, { fetchOrders } from '../order-history';

import { errorSelector, loadingSelector } from './selectors';

const OrdersAndSubscriptionsPage = () => {
  const { formatMessage } = useIntl();
  const dispatch = useDispatch();
  const isLoading = useSelector(loadingSelector);
  const hasError = useSelector(errorSelector);

  const isB2CSubsEnabled = (
    getConfig().ENABLE_B2C_SUBSCRIPTIONS?.toLowerCase() === 'true'
  );

  useEffect(() => {
    if (isB2CSubsEnabled) {
      document.title = 'Orders and Subscriptions | edX';
      sendTrackEvent('edx.bi.user.subscription.order-page.viewed');
      dispatch(fetchSubscriptions());
    }
    // TODO: We should fetch based on the route (ex: /orders/?orderPage=1)
    dispatch(fetchOrders(1));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const renderLoading = () => (
    <PageLoading
      srMessage={formatMessage({
        id: 'ecommerce.order.history.loading',
        defaultMessage: 'Loading orders and subscriptions...',
        description: 'Message when orders and subscriptions page is loading.',
      })}
    />
  );

  const renderOrdersandSubscriptions = () => (
    <>
      <Subscriptions />
      <OrderHistory isB2CSubsEnabled />
    </>
  );

  if (!isB2CSubsEnabled) {
    return (
      <div className="page__orders-and-subscriptions container-fluid container-mw-xl py-5">
        <OrderHistory isB2CSubsEnabled={false} />
      </div>
    );
  }

  return (
    <div className="page__orders-and-subscriptions container-fluid container-mw-xl py-4.5">
      <div className="section">
        <BasicAlert isVisible={hasError} />
        <FormattedMessage
          id="ecommerce.order.history.main.heading"
          defaultMessage="My orders and subscriptions"
          description="Heading for orders and subscriptions page."
        >
          {(text) => <h1 className="text-primary-700">{text}</h1>}
        </FormattedMessage>
        <FormattedMessage
          id="ecommerce.order.history.main.subtitle"
          defaultMessage="Manage your program subscriptions and view your order history."
          description="Subtitle of Heading for orders and subscriptions page."
        >
          {(text) => <span className="text-dark-900">{text}</span>}
        </FormattedMessage>
      </div>
      {isLoading ? renderLoading() : renderOrdersandSubscriptions()}
    </div>
  );
};

export default OrdersAndSubscriptionsPage;
