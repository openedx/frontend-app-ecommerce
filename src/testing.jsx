/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { IntlProvider } from '@edx/frontend-platform/i18n';

import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

const mockStore = configureMockStore();

jest.mock('@edx/frontend-platform/analytics', () => ({
  sendTrackEvent: jest.fn(),
  sendPageEvent: jest.fn(),
}));

const renderWithProviders = (component, store = null) => render(
  <IntlProvider locale="en">
    {store ? (
      <Provider store={mockStore(store)}>
        {component}
      </Provider>
    ) : (
      component
    )}
  </IntlProvider>,
);

// re-export everything
export * from '@testing-library/react';

// override render method
export { renderWithProviders as render };
