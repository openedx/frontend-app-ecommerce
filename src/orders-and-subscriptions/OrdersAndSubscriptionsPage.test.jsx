/* eslint-disable global-require */
import React from 'react';
import { render, screen } from '../testing';

import OrdersAndSubscriptionsPage from './OrdersAndSubscriptionsPage';

const storeMocks = require('../store/__mocks__/mockStore');
const emptyStoreMocks = require('../store/__mocks__/mockEmptyStore');

const {
  getByText,
  getByTestId,
  queryByText,
  queryByTestId,
} = screen;

const testHeadings = (hasSections = true) => {
  // Assert the main heading
  expect(getByText('My orders and subscriptions')).toBeInTheDocument();
  expect(
    getByText('Manage your program subscriptions and view your order history.'),
  ).toBeInTheDocument();

  if (hasSections) {
    // Assert Subscription and Order History sections are rendered
    expect(getByText('Subscriptions')).toBeInTheDocument();
    expect(getByText('Order History')).toBeInTheDocument();
  } else {
    // Assert Subscription and Order History sections are not rendered
    expect(queryByText('Subscriptions')).toBeNull();
    expect(queryByText('Order History')).toBeNull();
  }
};

describe('<OrdersAndSubscriptions />', () => {
  describe('Renders correctly in various states', () => {
    it('renders with orders and subscriptions', () => {
      render(<OrdersAndSubscriptionsPage />, storeMocks);
      testHeadings();

      // Assert alerts are not rendered
      expect(queryByTestId('basic-alert')).toBeNull();
    });

    it('renders alerts on errors', () => {
      const storeMocksWithErrors = {
        orderHistory: {
          ...emptyStoreMocks.orderHistory,
          loadingError: true,
        },
        subscriptions: {
          ...emptyStoreMocks.subscriptions,
          loadingError: true,
        },
      };

      render(<OrdersAndSubscriptionsPage />, storeMocksWithErrors);
      testHeadings();

      expect(getByTestId('basic-alert')).toBeInTheDocument();

      // Assert Subscription section renders empty state
      expect(queryByTestId('section-subscription-cards')).toBeNull();
      expect(getByTestId('section-subscription-upsell')).toBeInTheDocument();
    });

    it('renders with loading', () => {
      const storeMocksWithLoading = {
        orderHistory: {
          ...emptyStoreMocks.orderHistory,
          loading: true,
        },
        subscriptions: {
          ...emptyStoreMocks.subscriptions,
          loading: true,
        },
      };

      render(<OrdersAndSubscriptionsPage />, storeMocksWithLoading);
      testHeadings(false);

      // Assert loading message is rendered
      expect(getByText('Loading orders and subscriptions...'))
        .toBeInTheDocument();

      // Assert alerts are not rendered
      expect(queryByTestId('basic-alert')).toBeNull();
    });
  });
});
