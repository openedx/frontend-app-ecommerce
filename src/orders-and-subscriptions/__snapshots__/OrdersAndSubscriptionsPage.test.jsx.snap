// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrdersAndSubscriptions /> Renders correctly in various states renders with orders and subscriptions 1`] = `
<div
  className="page__orders-and-subscriptions container-fluid py-4.5"
>
  <div
    className="section"
  >
    <h1
      className="text-primary-700"
    >
      My orders and subscriptions
    </h1>
    <span
      className="text-dark-900"
    >
      Manage your program subscriptions and view your order history.
    </span>
  </div>
  <section>
    <h2>
      Subscriptions
    </h2>
    <div
      className="section flex-md-row align-items-start align-items-md-center justify-content-between"
    >
      <span
        className="text-dark-900"
      >
        To view your receipts, change your payment method or cancel your subscription, click 
        <i>
          Manage my subscription
        </i>
        .
      </span>
      <button
        className="text-nowrap btn btn-primary btn-sm"
        disabled={false}
        type="button"
      >
        Manage my subscription
        <span
          className="pgn__icon btn-icon-after pgn__icon__sm"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 19H5V5h7V3H3v18h18v-9h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      aria-controls="scrollbar"
      aria-valuemax={100}
      aria-valuemin={0}
      aria-valuenow={0}
      className="pgn__scrollable-body subscription-scrollable mx-n1 pgn__scrollable-body-scroll-top pgn__scrollable-body-scroll-bottom"
      onScroll={[Function]}
      role="scrollbar"
      tabIndex="0"
    >
      <div />
      <div
        className="pgn__scrollable-body-content"
      >
        <div
          className="section section-gap-lg mx-1"
        >
          <div
            className="bg-light-200 p-3 pgn__card card"
            tabIndex="-1"
          >
            <div
              className="section flex-column-reverse flex-sm-row align-items-start align-items-sm-center"
            >
              <h3
                className="text-info-500 m-0"
              >
                Blockchain Fundamentals
              </h3>
              <span
                className="text-capitalize badge badge-light"
              >
                trial
              </span>
            </div>
            <p
              className="small text-gray-500 m-0"
            >
              University of California, Berkeley
            </p>
          </div>
          <div
            className="bg-light-200 p-3 pgn__card card"
            tabIndex="-1"
          >
            <div
              className="section flex-column-reverse flex-sm-row align-items-start align-items-sm-center"
            >
              <h3
                className="text-info-500 m-0"
              >
                Critical Thinking
              </h3>
              <span
                className="text-capitalize badge badge-light"
              >
                active
              </span>
            </div>
            <p
              className="small text-gray-500 m-0"
            >
              Simmons University
            </p>
          </div>
          <div
            className="bg-light-200 p-3 pgn__card card"
            tabIndex="-1"
          >
            <div
              className="section flex-column-reverse flex-sm-row align-items-start align-items-sm-center"
            >
              <h3
                className="text-info-500 m-0"
              >
                Blockchain Fundamentals
              </h3>
              <span
                className="text-capitalize badge badge-light"
              >
                inactive
              </span>
            </div>
            <p
              className="small text-gray-500 m-0"
            >
              University of California, Berkeley
            </p>
          </div>
          <div
            className="bg-light-200 p-3 pgn__card card"
            tabIndex="-1"
          >
            <div
              className="section flex-column-reverse flex-sm-row align-items-start align-items-sm-center"
            >
              <h3
                className="text-info-500 m-0"
              >
                Critical Thinking
              </h3>
              <span
                className="text-capitalize badge badge-light"
              >
                inactive
              </span>
            </div>
            <p
              className="small text-gray-500 m-0"
            >
              Simmons University
            </p>
          </div>
        </div>
      </div>
      <div />
    </div>
  </section>
  <section
    className="page__order-history"
  >
    <h2>
      Order History
    </h2>
    <div>
      <nav
        aria-label="pagination navigation"
        className="pagination-default"
      >
        <div
          aria-atomic={true}
          aria-live="polite"
          aria-relevant="text"
          className="sr-only"
        >
          Page 2, Current Page, of 10
        </div>
        <ul
          className="pagination"
        >
          <li
            className="page-item"
          >
            <button
              aria-label="Previous, Page 1"
              className="previous page-link btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <div>
                <span
                  className="pgn__icon"
                >
                  <svg
                    aria-hidden={true}
                    fill="none"
                    focusable={false}
                    height={24}
                    role="img"
                    viewBox="0 0 24 24"
                    width={24}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12l4.58-4.59z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
                Previous
              </div>
            </button>
          </li>
          <li
            className="page-item"
          >
            <button
              aria-label="Next, Page 3"
              className="next page-link btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <div>
                Next
                <span
                  className="pgn__icon"
                >
                  <svg
                    aria-hidden={true}
                    fill="none"
                    focusable={false}
                    height={24}
                    role="img"
                    viewBox="0 0 24 24"
                    width={24}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6-6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </div>
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </section>
</div>
`;
